File: analyzers/lang_utils.py
"""Language-aware pattern matching for tone & interaction analysis.

Provides:
- Language detection from message text
- Per-language pattern tables for English (en) and Indonesian (id)
- Universal structural/character-level heuristics that work across languages
- A match function that selects the right patterns based on detected language
"""

import re
from typing import Optional

# ── Language detection ────────────────────────────────────────────────


def detect_language(texts: list[str]) -> str:
    """Detect the dominant language from a collection of message texts.

    Falls back to 'en' if langdetect is not installed or detection fails.
    """
    try:
        from langdetect import detect, DetectorFactory

        DetectorFactory.seed = 0

        sample = " ".join(texts[:100])
        if len(sample.strip()) < 10:
            return "en"
        return detect(sample)
    except ImportError:
        return "en"
    except Exception:
        return "en"


# ── Regex-based tone patterns (word-boundary aware) ───────────────────
# These are used by analyzers/tone.py via re.search()

EN_TONE = {
    "apology": [
        r"\bsorry\b", r"\bapologi[sz]e\b", r"\bmy\s+mistake\b", r"\bmy\s+fault\b",
        r"\bI\s+was\s+wrong\b", r"\bI\s+apologi[sz]e\b", r"\bmy\s+bad\b",
        r"\bthat\s+was\s+incorrect\b", r"\bI\s+messed\s+up\b",
        r"\byou\s+are\s+right\b", r"\byou're\s+right\b", r"\bgood\s+catch\b",
    ],
    "confidence": [
        r"\bdefinitely\b", r"\bcertainly\b", r"\bI\s+am\s+sure\b", r"\bI'm\s+sure\b",
        r"\bclearly\b", r"\bobviously\b", r"\bwithout\s+doubt\b", r"\babsolutely\b",
        r"\bindeed\b", r"\bexactly\b", r"\bprecisely\b", r"\bthat\s+is\s+correct\b",
        r"\bthat's\s+correct\b", r"\bno\s+problem\b",
    ],
    "uncertainty": [
        r"\bmaybe\b", r"\bperhaps\b", r"\bI\s+think\b", r"\bnot\s+sure\b",
        r"\bmight\s+be\b", r"\bcould\s+be\b", r"\bpossibly\b", r"\bprobably\b",
        r"\bI\s+believe\b", r"\bit\s+seems\b", r"\bappears?\b", r"\bI\s+guess\b",
        r"\bunsure\b", r"\bunclear\b", r"\bI\s+don't\s+know\b", r"\bI\s+dont\s+know\b",
        r"\bI\s+would\s+guess\b", r"\bsounds?\s+like\b",
    ],
    "helpfulness": [
        r"\blet\s+me\b", r"\bI\s+can\b", r"\bI'll\b", r"\bI\s+will\b",
        r"\bhere's\b", r"\bhere\s+is\b", r"\byou\s+can\b", r"\bwould\s+you\s+like\b",
        r"\blet's\b", r"\bI\s+recommend\b", r"\bI\s+suggest\b", r"\bfeel\s+free\b",
        r"\bhappy\s+to\b", r"\bglad\s+to\b", r"\byou're\s+welcome\b",
        r"\bI\s+will\s+help\b", r"\bI\s+can\s+help\b", r"\bhere\s+you\s+go\b",
        r"\bdid\s+that\s+help\b", r"\bdoes\s+that\s+help\b",
    ],
    "self_correction": [
        r"\bactually\b", r"\bwait\b", r"\blet\s+me\s+reconsider\b",
        r"\bon\s+second\s+thought\b", r"\brather\b", r"\bI\s+mean\b",
        r"\bscratch\s+that\b", r"\blet\s+me\s+rephrase\b", r"\binstead\b",
        r"\bcorrection\b", r"\bI\s+should\s+have\b",
        r"\bthat's\s+not\s+right\b", r"\blet\s+me\s+fix\b", r"\bhold\s+on\b",
        r"\bI\s+misread\b", r"\bI\s+misunderstood\b",
    ],
    "agent_question": [
        r"\bwhat\b", r"\bhow\b", r"\bwhy\b", r"\bwhen\b", r"\bwhere\b",
        r"\bwho\b", r"\bwhich\b", r"\bcan\s+you\b", r"\bcould\s+you\b",
        r"\bwould\s+you\b", r"\bdo\s+you\b", r"\bdoes\s+that\b",
        r"\bis\s+that\b", r"\bare\s+you\b", r"\bshall\s+I\b",
        r"\bwould\s+you\s+like\b", r"\bdo\s+you\s+want\b",
    ],
}

ID_TONE = {
    "apology": [
        r"\bmaaf\b", r"\bmaafkan\b", r"\bminta\s+maaf\b",
        r"\bsaya\s+salah\b", r"\baku\s+salah\b",
        r"\bkesalahan\w*\b", r"\bsalah\s+saya\b",
        r"\bseharusnya\b", r"\bsepertinya\s+salah\b",
        r"\bmaaf\s+ya\b", r"\boh\s+maaf\b",
    ],
    "confidence": [
        r"\btentu\b", r"\btentunya\b", r"\bpasti\b", r"\byakin\b",
        r"\bjelas\b", r"\b100%\b", r"\bsudah\s+pasti\b",
        r"\bdengan\s+yakin\b", r"\bgampang\b", r"\bmudah\b",
        r"\bsaya\s+yakin\b", r"\baku\s+yakin\b",
        r"\btidak\s+ada\s+masalah\b",
    ],
    "uncertainty": [
        r"\bmungkin\b", r"\bbarangkali\b", r"\bsepertinya\b", r"\bkayaknya\b",
        r"\bkurang\s+yakin\b", r"\bbelum\s+tahu\b", r"\btidak\s+tahu\b",
        r"\bgak\s+tahu\b", r"\bnggak\s+tahu\b", r"\bragu\b",
        r"\bkurang\s+tau\b", r"\bbisa\s+jadi\b", r"\btampaknya\b",
        r"\baku\s+pikir\b", r"\bsaya\s+pikir\b",
    ],
    "helpfulness": [
        r"\bbiar\s+saya\b", r"\bsaya\s+bisa\b", r"\baku\s+bisa\b",
        r"\bsaya\s+akan\b", r"\baku\s+akan\b", r"\bberikut\b",
        r"\bini\s+dia\b", r"\bkamu\s+bisa\b", r"\banda\s+bisa\b",
        r"\bmari\b", r"\bsaya\s+sarankan\b", r"\bsaya\s+rekomendasikan\b",
        r"\bsilakan\b", r"\bsilahkan\b", r"\bsenang\s+membantu\b",
        r"\bdengan\s+senang\s+hati\b", r"\bsama-sama\b",
        r"\bsaya\s+akan\s+membantu\b", r"\baku\s+akan\s+membantu\b",
        r"\bapakah\s+membantu\b",
    ],
    "self_correction": [
        r"\bsebenarnya\b", r"\btunggu\b", r"\bentar\b",
        r"\bmaksud\w*\s+saya\b", r"\blupakan\b",
        r"\bbiar\s+saya\s+ulang\b", r"\bganti\b",
        r"\boh\s+iya\b", r"\bmaksudku\b", r"\bmaksud\s+saya\b",
        r"\bbiar\s+saya\s+perbaiki\b", r"\bsaya\s+keliru\b",
        r"\bsaya\s+salah\s+baca\b", r"\bbukan\s+begitu\b",
    ],
    "agent_question": [
        r"\bapa\b", r"\bbagaimana\b", r"\bmengapa\b", r"\bkenapa\b",
        r"\bkapan\b", r"\bdimana\b", r"\bsiapa\b", r"\byang\s+mana\b",
        r"\bbisakah\b", r"\bbisa\s+kamu\b", r"\bbisa\s+anda\b",
        r"\bapakah\b", r"\bapakah\s+itu\b",
        r"\bapakah\s+kamu\b", r"\bapakah\s+anda\b",
        r"\bbingung\b", r"\bmaukah\b",
    ],
}

# ── Simple substring interaction patterns (used by analyzers/interaction.py) ─

EN_INTERACTION = {
    "correction": [
        "wrong", "nope", "incorrect", "actually", "i meant",
        "what i meant", "that's not", "don't do", "do not",
        "shouldn't", "should not", "try again", "redo", "re-do",
        "not yet", "not what i", "not correct", "not working",
        "doesn't work", "didn't work", "still broken", "not right",
    ],
    "exit_positive": [
        "thanks", "thank", "perfect", "great", "goodbye", "bye", "done",
        "that's all", "all good", "resolved", "/exit",
    ],
    "clarification_question": [
        "what do you mean", "can you clarify", "could you clarify",
        "do you mean", "are you saying", "do you want me to",
        "would you like me to", "to confirm", "just to be clear",
        "let me make sure", "so you want", "is this what you",
        "am i understanding", "to clarify", "which one",
        "did you mean", "let me know if",
    ],
}

ID_INTERACTION = {
    "correction": [
        "salah", "bukan", "bukan itu", "maksudnya", "seharusnya",
        "bukan begitu", "jangan", "coba lagi", "ulang", "ulangi",
        "belum", "bukan ini", "salah semua", "masih salah",
        "tidak benar", "nggak", "gak", "tidak sesuai",
        "bukan gitu", "salah lagi",
    ],
    "exit_positive": [
        "makasih", "terima kasih", "terimakasih", "thanks", "thank",
        "sempurna", "bagus", "hebat", "mantap", "oke", "ok",
        "selesai", "sudah", "cukup", "done", "bye", "dadah",
        "baik terima kasih", "baiklah",
    ],
    "clarification_question": [
        "maksud kamu", "maksud anda", "apa yang kamu maksud",
        "bisa dijelaskan", "bisa klarifikasi", "apakah kamu mengatakan",
        "apakah maksudmu", "apakah anda mengatakan",
        "apakah kamu mau", "untuk memastikan", "biar jelas",
        "jadi kamu mau", "apakah ini yang", "apakah saya memahami",
        "untuk klarifikasi", "yang mana", "tolong jelaskan",
    ],
}

# ── Pattern selection ─────────────────────────────────────────────────

LANGUAGE_TONE_PATTERNS: dict[str, dict[str, list[str]]] = {
    "en": EN_TONE,
    "id": ID_TONE,
    "ms": ID_TONE,  # Malay shares most patterns with Indonesian
}

LANGUAGE_INTERACTION_PATTERNS: dict[str, dict[str, list[str]]] = {
    "en": EN_INTERACTION,
    "id": ID_INTERACTION,
    "ms": ID_INTERACTION,
}


def get_tone_patterns(language: str) -> dict[str, list[str]]:
    """Return regex patterns for tone analysis in the given language.

    Falls back to English if the language is unsupported.
    """
    return LANGUAGE_TONE_PATTERNS.get(language, EN_TONE)


def get_interaction_patterns(language: str) -> dict[str, list[str]]:
    """Return substring patterns for interaction analysis in the given language.

    Falls back to English if the language is unsupported.
    """
    return LANGUAGE_INTERACTION_PATTERNS.get(language, EN_INTERACTION)


# ── Universal structural / character-level heuristics ─────────────────
# These work across languages regardless of vocabulary.

# Emoji that carry tone/interaction signals
TONE_EMOJI = {
    "apology": {"😅", "😓", "😔", "🙏", "😬", "🥺"},
    "confidence": {"💪", "🔥", "✅", "✔️", "🎯"},
    "uncertainty": {"🤔", "😕", "🤷", "🤷‍♂️", "🤷‍♀️", "🧐"},
    "helpfulness": {"👍", "👌", "🙌", "✨", "💡", "😊"},
    "self_correction": {"🔄", "✏️", "📝"},
    "positive": {"😊", "👍", "🎉", "✅", "🙌", "💯", "🔥", "✨"},
    "negative": {"😞", "😠", "😤", "👎", "❌", "💀"},
    "question": {"❓", "❔", "🤔"},
    "gratitude": {"🙏", "🙏🏻", "🙏🏼", "😊"},
}


def has_question_mark(text: str) -> bool:
    """Universal: message contains a question mark."""
    return "?" in text


def has_exclamation(text: str) -> bool:
    """Universal: message contains an exclamation mark."""
    return "!" in text


def has_repeated_chars(text: str) -> bool:
    """Universal: detect repeated characters signalling emotion (e.g. 'noooo', 'whaat')."""
    return bool(re.search(r"(.)\1{3,}", text))


def has_emoji(text: str, category: str) -> bool:
    """Check if the text contains any emoji from the given category."""
    emojis = TONE_EMOJI.get(category, set())
    for ch in text:
        if ch in emojis:
            return True
    # Also check for multi-char emoji sequences
    for emo in emojis:
        if len(emo) > 1 and emo in text:
            return True
    return False


def count_all_caps_words(text: str) -> int:
    """Count words that are entirely uppercase (≥2 chars), excluding the first word of a sentence."""
    words = text.split()
    count = 0
    for i, w in enumerate(words):
        w_clean = w.strip(".!?\"'(),;:")
        if len(w_clean) >= 2 and w_clean.isupper():
            # Skip the very first word (could be 'I' or start of sentence)
            if i == 0 and len(w_clean) == 1:
                continue
            count += 1
    return count


# ── Convenience matcher ───────────────────────────────────────────────

def match_tone_pattern(patterns: dict[str, list[str]], category: str, text: str) -> bool:
    """Check if *any* regex pattern in the given category matches the text."""
    pats = patterns.get(category, [])
    for p in pats:
        if re.search(p, text, re.IGNORECASE):
            return True
    return False


def match_interaction_pattern(patterns: dict[str, list[str]], category: str, text: str) -> bool:
    """Check if *any* substring pattern in the given category appears in the text."""
    pats = patterns.get(category, [])
    for p in pats:
        if p in text:
            return True
    return False

File: src/analyzers/lang_utils.ts
import { franc } from "franc";

const EN_TONE = {
  apology: [
    "\\bsorry\\b",
    "\\bapologi[sz]e\\b",
    "\\bmy\\s+mistake\\b",
    "\\bmy\\s+fault\\b",
    "\\bI\\s+was\\s+wrong\\b",
    "\\bI\\s+apologi[sz]e\\b",
    "\\bmy\\s+bad\\b",
    "\\bthat\\s+was\\s+incorrect\\b",
    "\\bI\\s+messed\\s+up\\b",
    "\\byou\\s+are\\s+right\\b",
    "\\byou\u2019re\\s+right\\b",
    "\\bgood\\s+catch\\b",
  ],
  confidence: [
    "\\bdefinitely\\b",
    "\\bcertainly\\b",
    "\\bI\\s+am\\s+sure\\b",
    "\\bI\u2019m\\s+sure\\b",
    "\\bclearly\\b",
    "\\bobviously\\b",
    "\\bwithout\\s+doubt\\b",
    "\\babsolutely\\b",
    "\\bindeed\\b",
    "\\bexactly\\b",
    "\\bprecisely\\b",
    "\\bthat\\s+is\\s+correct\\b",
    "\\bthat\u2019s\\s+correct\\b",
    "\\bno\\s+problem\\b",
  ],
  uncertainty: [
    "\\bmaybe\\b",
    "\\bperhaps\\b",
    "\\bI\\s+think\\b",
    "\\bnot\\s+sure\\b",
    "\\bmight\\s+be\\b",
    "\\bcould\\s+be\\b",
    "\\bpossibly\\b",
    "\\bprobably\\b",
    "\\bI\\s+believe\\b",
    "\\bit\\s+seems\\b",
    "\\bappears?\\b",
    "\\bI\\s+guess\\b",
    "\\bunsure\\b",
    "\\bunclear\\b",
    "\\bI\\s+don\u2019t\\s+know\\b",
    "\\bI\\s+would\\s+guess\\b",
    "\\bsounds?\\s+like\\b",
  ],
  helpfulness: [
    "\\blet\\s+me\\b",
    "\\bI\\s+can\\b",
    "\\bI\u2019ll\\b",
    "\\bI\\s+will\\b",
    "\\bhere\u2019s\\b",
    "\\bhere\\s+is\\b",
    "\\byou\\s+can\\b",
    "\\bwould\\s+you\\s+like\\b",
    "\\blet\u2019s\\b",
    "\\bI\\s+recommend\\b",
    "\\bI\\s+suggest\\b",
    "\\bfeel\\s+free\\b",
    "\\bhappy\\s+to\\b",
    "\\bglad\\s+to\\b",
    "\\byou\u2019re\\s+welcome\\b",
    "\\bI\\s+will\\s+help\\b",
    "\\bI\\s+can\\s+help\\b",
    "\\bhere\\s+you\\s+go\\b",
    "\\bdid\\s+that\\s+help\\b",
    "\\bdoes\\s+that\\s+help\\b",
  ],
  self_correction: [
    "\\bactually\\b",
    "\\bwait\\b",
    "\\blet\\s+me\\s+reconsider\\b",
    "\\bon\\s+second\\s+thought\\b",
    "\\brather\\b",
    "\\bI\\s+mean\\b",
    "\\bscratch\\s+that\\b",
    "\\blet\\s+me\\s+rephrase\\b",
    "\\binstead\\b",
    "\\bcorrection\\b",
    "\\bI\\s+should\\s+have\\b",
    "\\bthat\u2019s\\s+not\\s+right\\b",
    "\\blet\\s+me\\s+fix\\b",
    "\\bhold\\s+on\\b",
    "\\bI\\s+misread\\b",
    "\\bI\\s+misunderstood\\b",
  ],
  agent_question: [
    "\\bwhat\\b",
    "\\bhow\\b",
    "\\bwhy\\b",
    "\\bwhen\\b",
    "\\bwhere\\b",
    "\\bwho\\b",
    "\\bwhich\\b",
    "\\bcan\\s+you\\b",
    "\\bcould\\s+you\\b",
    "\\bwould\\s+you\\b",
    "\\bdo\\s+you\\b",
    "\\bdoes\\s+that\\b",
    "\\bis\\s+that\\b",
    "\\bare\\s+you\\b",
    "\\bshall\\s+I\\b",
    "\\bwould\\s+you\\s+like\\b",
    "\\bdo\\s+you\\s+want\\b",
  ],
};

const ID_TONE = {
  apology: [
    "\\bmaaf\\b",
    "\\bmaafkan\\b",
    "\\bminta\\s+maaf\\b",
    "\\bsaya\\s+salah\\b",
    "\\baku\\s+salah\\b",
    "\\bkesalahan\\w*\\b",
    "\\bsalah\\s+saya\\b",
    "\\bseharusnya\\b",
    "\\bsepertinya\\s+salah\\b",
    "\\bmaaf\\s+ya\\b",
    "\\boh\\s+maaf\\b",
  ],
  confidence: [
    "\\btentu\\b",
    "\\btentunya\\b",
    "\\bpasti\\b",
    "\\byakin\\b",
    "\\bjelas\\b",
    "\\b100%\\b",
    "\\bsudah\\s+pasti\\b",
    "\\bdengan\\s+yakin\\b",
    "\\bgampang\\b",
    "\\bmudah\\b",
    "\\bsaya\\s+yakin\\b",
    "\\baku\\s+yakin\\b",
    "\\btidak\\s+ada\\s+masalah\\b",
  ],
  uncertainty: [
    "\\bmungkin\\b",
    "\\bbarangkali\\b",
    "\\bsepertinya\\b",
    "\\bkayaknya\\b",
    "\\bkurang\\s+yakin\\b",
    "\\bbelum\\s+tahu\\b",
    "\\btidak\\s+tahu\\b",
    "\\bgak\\s+tahu\\b",
    "\\bnggak\\s+tahu\\b",
    "\\bragu\\b",
    "\\bkurang\\s+tau\\b",
    "\\bbisa\\s+jadi\\b",
    "\\btampaknya\\b",
    "\\baku\\s+pikir\\b",
    "\\bsaya\\s+pikir\\b",
  ],
  helpfulness: [
    "\\bbiar\\s+saya\\b",
    "\\bsaya\\s+bisa\\b",
    "\\baku\\s+bisa\\b",
    "\\bsaya\\s+akan\\b",
    "\\baku\\s+akan\\b",
    "\\bberikut\\b",
    "\\bini\\s+dia\\b",
    "\\bkamu\\s+bisa\\b",
    "\\banda\\s+bisa\\b",
    "\\bmari\\b",
    "\\bsaya\\s+sarankan\\b",
    "\\bsaya\\s+rekomendasikan\\b",
    "\\bsilakan\\b",
    "\\bsilahkan\\b",
    "\\bsenang\\s+membantu\\b",
    "\\bdengan\\s+senang\\s+hati\\b",
    "\\bsama-sama\\b",
    "\\bsaya\\s+akan\\s+membantu\\b",
    "\\baku\\s+akan\\s+membantu\\b",
    "\\bapakah\\s+membantu\\b",
  ],
  self_correction: [
    "\\bsebenarnya\\b",
    "\\btunggu\\b",
    "\\bentar\\b",
    "\\bmaksud\\w*\\s+saya\\b",
    "\\blupakan\\b",
    "\\bbiar\\s+saya\\s+ulang\\b",
    "\\bganti\\b",
    "\\boh\\s+iya\\b",
    "\\bmaksudku\\b",
    "\\bmaksud\\s+saya\\b",
    "\\bbiar\\s+saya\\s+perbaiki\\b",
    "\\bsaya\\s+keliru\\b",
    "\\bsaya\\s+salah\\s+baca\\b",
    "\\bbukan\\s+begitu\\b",
  ],
  agent_question: [
    "\\bapa\\b",
    "\\bbagaimana\\b",
    "\\bmengapa\\b",
    "\\bkenapa\\b",
    "\\bkapan\\b",
    "\\bdimana\\b",
    "\\bsiapa\\b",
    "\\byang\\s+mana\\b",
    "\\bbisakah\\b",
    "\\bbisa\\s+kamu\\b",
    "\\bbisa\\s+anda\\b",
    "\\bapakah\\b",
    "\\bapakah\\s+itu\\b",
    "\\bapakah\\s+kamu\\b",
    "\\bapakah\\s+anda\\b",
    "\\bbingung\\b",
    "\\bmaukah\\b",
  ],
};


const EN_INTERACTION = {
  correction: [
    "wrong",
    "nope",
    "incorrect",
    "i meant",
    "what i meant",
    "that\u2019s not",
    "shouldn\u2019t",
    "should not",
    "try again",
    "redo",
    "re-do",
    "not yet",
    "not what i",
    "not correct",
    "not working",
    "doesn\u2019t work",
    "didn\u2019t work",
    "still broken",
    "not right",
  ],
  exit_positive: [
    "thanks",
    "thank",
    "perfect",
    "great",
    "goodbye",
    "bye",
    "done",
    "that\u2019s all",
    "all good",
    "resolved",
    "/exit",
  ],
  clarification_question: [
    "what do you mean",
    "can you clarify",
    "could you clarify",
    "do you mean",
    "are you saying",
    "do you want me to",
    "would you like me to",
    "to confirm",
    "just to be clear",
    "let me make sure",
    "so you want",
    "is this what you",
    "am i understanding",
    "to clarify",
    "which one",
    "did you mean",
    "let me know if",
  ],
};

const ID_INTERACTION = {
  correction: [
    "salah",
    "bukan",
    "bukan itu",
    "maksudnya",
    "seharusnya",
    "bukan begitu",
    "coba lagi",
    "ulang",
    "ulangi",
    "belum",
    "bukan ini",
    "salah semua",
    "masih salah",
    "tidak benar",
    "nggak",
    "gak",
    "tidak sesuai",
    "bukan gitu",
    "salah lagi",
  ],
  exit_positive: [
    "makasih",
    "terima kasih",
    "terimakasih",
    "thanks",
    "thank",
    "sempurna",
    "bagus",
    "hebat",
    "mantap",
    "oke",
    "ok",
    "selesai",
    "sudah",
    "cukup",
    "done",
    "bye",
    "dadah",
    "baik terima kasih",
    "baiklah",
  ],
  clarification_question: [
    "maksud kamu",
    "maksud anda",
    "apa yang kamu maksud",
    "bisa dijelaskan",
    "bisa klarifikasi",
    "apakah kamu mengatakan",
    "apakah maksudmu",
    "apakah anda mengatakan",
    "apakah kamu mau",
    "untuk memastikan",
    "biar jelas",
    "jadi kamu mau",
    "apakah ini yang",
    "apakah saya memahami",
    "untuk klarifikasi",
    "yang mana",
    "tolong jelaskan",
  ],
};

export const STRONG_CORRECTION_PATTERNS_EN = [
  "wrong", "nope", "incorrect",
  "try again", "redo", "re-do", "not correct", "not working",
  "doesn't work", "didn't work", "still broken", "not right",
];

export const WEAK_CORRECTION_PATTERNS_EN = [
  "i meant", "what i meant", "that's not",
  "shouldn't", "should not", "not yet", "not what i",
];

export const STRONG_CORRECTION_PATTERNS_ID = [
  "salah", "bukan itu", "bukan begitu",
  "coba lagi", "ulang", "ulangi", "salah semua",
  "masih salah", "tidak benar", "bukan ini", "salah lagi",
];

export const WEAK_CORRECTION_PATTERNS_ID = [
  "bukan", "maksudnya", "seharusnya", "belum",
  "nggak", "gak", "tidak sesuai", "bukan gitu",
];


const TONE_EMOJI = {
  apology: new Set(["😅", "😓", "😔", "🙏", "😬", "🥺"]),
  confidence: new Set(["💪", "🔥", "✅", "✔️", "🎯"]),
  uncertainty: new Set(["🤔", "😕", "🤷", "🧐"]),
  helpfulness: new Set(["👍", "👌", "🙌", "✨", "💡", "😊"]),
  self_correction: new Set(["🔄", "✏️", "📝"]),
  positive: new Set(["😊", "👍", "🎉", "✅", "🙌", "💯", "🔥", "✨"]),
  negative: new Set(["😞", "😠", "😤", "👎", "❌", "💀"]),
  question: new Set(["❓", "❔", "🤔"]),
  gratitude: new Set(["🙏", "😊"]),
};

type ToneCategory =
  | "apology"
  | "confidence"
  | "uncertainty"
  | "helpfulness"
  | "self_correction"
  | "agent_question";

type InteractionCategory = "correction" | "exit_positive" | "clarification_question";

type TonePatterns = Record<ToneCategory, string[]>;
type InteractionPatterns = Record<InteractionCategory, string[]>;

function getToneTable(lang: string): TonePatterns {
  switch (lang) {
    case "id":
      return ID_TONE;
    default:
      return EN_TONE;
  }
}

function getInteractionTable(lang: string): InteractionPatterns {
  switch (lang) {
    case "id":
      return ID_INTERACTION;
    default:
      return EN_INTERACTION;
  }
}

const ISO_TO_SHORT: Record<string, string> = {
  eng: "en",
  ind: "id",
};

export function detectLanguage(texts: string[]): string {
  const sample = texts.slice(0, 100).join(" ").trim();
  if (sample.length < 10) return "en";
  const detected = franc(sample);
  return ISO_TO_SHORT[detected] ?? "en";
}

export function getTonePatterns(lang: string): TonePatterns {
  return getToneTable(lang);
}

export function getInteractionPatterns(lang: string): InteractionPatterns {
  return getInteractionTable(lang);
}

export function matchTonePattern(
  patterns: TonePatterns,
  category: ToneCategory,
  text: string,
): boolean {
  const regexes = patterns[category] ?? [];
  return regexes.some((re) => new RegExp(re, "i").test(text));
}

export function scoreInteractionPattern(
  patterns: InteractionPatterns,
  category: InteractionCategory,
  text: string,
  sentimentScore: number = 0,
  lang: string = "en",
): number {
  const substrings = patterns[category] ?? [];
  const lower = text.toLowerCase();
  const matched = substrings.filter((sub) => lower.includes(sub.toLowerCase()));

  // exit_positive: binary
  if (category === "exit_positive") {
    return matched.length > 0 ? 1 : 0;
  }

  // clarification_question: stepped (no modifiers)
  if (category === "clarification_question") {
    if (matched.length === 0) return 0;
    if (matched.length === 1) return 0.4;
    return 0.6;
  }

  // correction: multi-signal scoring
  if (matched.length === 0) return 0;

  // Step 1: pattern density
  let score: number;
  if (matched.length === 1) score = 0.3;
  else if (matched.length === 2) score = 0.5;
  else score = 0.7;

  // Step 2: strength boost
  const strongPatterns = lang === "id" ? STRONG_CORRECTION_PATTERNS_ID : STRONG_CORRECTION_PATTERNS_EN;
  const hasStrong = matched.some((m) =>
    strongPatterns.some((sp) => m.toLowerCase().includes(sp.toLowerCase()))
  );
  if (hasStrong) score = Math.min(1, score + 0.2);

  // Step 3: sentiment modifier (only if score > 0)
  if (score > 0) {
    if (sentimentScore < 0) score *= 1.2;
    else if (sentimentScore > 0.3) score *= 0.8;
  }

  // Step 4: length normalization
  const wordCount = text.split(/\s+/).filter(Boolean).length;
  if (score > 0) {
    if (wordCount <= 5) score += 0.1;
    else if (wordCount > 100) score -= 0.1;
  }

  // Step 5: clamp
  return Math.min(1, Math.max(0, score));
}

export function matchInteractionPattern(
  patterns: InteractionPatterns,
  category: InteractionCategory,
  text: string,
): number {
  return scoreInteractionPattern(patterns, category, text, 0, "en");
}

export function hasInteractionPattern(
  patterns: InteractionPatterns,
  category: InteractionCategory,
  text: string,
): boolean {
  return scoreInteractionPattern(patterns, category, text) > 0;
}

export function hasQuestionMark(text: string): boolean {
  return text.includes("?");
}

export function hasExclamation(text: string): boolean {
  return text.includes("!");
}

export function hasRepeatedChars(text: string): boolean {
  return /(.)\1{2,}/.test(text);
}

export function hasEmoji(
  text: string,
  category: keyof typeof TONE_EMOJI,
): boolean {
  const emojis = TONE_EMOJI[category];
  if (!emojis) return false;
  return Array.from(emojis).some((emoji) => text.includes(emoji));
}

export function countAllCapsWords(text: string): number {
  const words = text.match(/\b[A-Z]{2,}\b/g);
  return words?.length ?? 0;
}
